### DISCUSSION TITLE
Trouble in npm monorepo

### DISCUSSION THREAD

**Discussion by nephix** [expo/router#235] [Q&A]
### Summary

I'm having trouble adding expo router to our npm monorepo.

I followed the steps on https://expo.github.io/router/docs/ and added some of the guidelines from https://docs.expo.dev/guides/monorepos/

and

```
npx expo start --ios
```

always ends up with:

```
iOS Bundling failed 27ms
Unable to resolve "@expo/metro-runtime" from "node_modules/expo-router/entry.js"
```

Locally, I can see that `@expo/metro-runtime` indeed does not exist in

> node_modules
> apps/expo-app/node_modules

even though it's listed as a dependency in

> apps/expo-app/node_modules/expo-router/package.json

After installing this "missing" package (and a bunch of others that cant be found) I run into a dead end and get:

```
 ERROR  TypeError: (0, _$$_REQUIRE(_dependencyMap[4], "react/jsx-runtime").jsx) is not a function. (In '(0, _$$_REQUIRE(_dependencyMap[4], "react/jsx-runtime").jsx)(ContextNavigator, {
          context: context
        })', '(0, _$$_REQUIRE(_dependencyMap[4], "react/jsx-runtime").jsx)' is undefined)

This error is located at:
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent)

```

**EDIT**:

After upgrading to Expo SDK 48, I'm getting now:

```
 ERROR  TypeError: undefined is not a function

This error is located at:
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes
 ERROR  TypeError: undefined is not a function
```

![CleanShot 2023-02-26 at 09 08 28](https://user-images.githubusercontent.com/3169205/221401697-38e94cc4-a0dc-4380-8937-a542bea5fe86.png)

### Minimal reproducible example

This is a very minimal example monorepo, it only contains the expo app w/ expo-router

```
git clone git@github.com:nephix/expo-router-repro.git
cd expo-router-repro
npm install
cd apps/expo-app
npx expo start --ios
```

**Comment by EvanBacon** [comment]
`react/jsx-runtime` appears to not have a function named `jsx`. This could mean you have a really old version of React being imported somewhere.

**Reply by nephix** [reply]
Weird I only installed expo via a template in the minimal example.

There is literally nothing else in the monorepo

**Reply by nephix** [reply]
I updated it now to Expo SDK 48 and the error now is:

```
 ERROR  TypeError: undefined is not a function

This error is located at:
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes
 ERROR  TypeError: undefined is not a function
```

![CleanShot 2023-02-26 at 09 08 28](https://user-images.githubusercontent.com/3169205/221401718-a641ab49-e7e0-4d14-813c-401755da97bd.png)
